pkg/kv/kvserver/replica_init.go

```diff
@@ -127,9 +181,23 @@ func newUnloadedReplica(
 		r.loadStats = load.NewReplicaLoad(store.Clock(), store.cfg.StorePool.GetNodeLocalityString)
 	}
 
-	// NB: the state will be loaded when the replica gets initialized.
-	r.mu.state = uninitState
-	r.rangeStr.store(replicaID, uninitState.Desc)
+	r.mu.state = state
+	// Only do this if there was a previous lease. This shouldn't be important
```

Comment:
Right, I similarly convinced myself that it was ok to lift this, but similarly felt that it's slightly out of place (just like other stuff in this func that is needed only for initialized replicas, e.g. closed timestamp inits etc). Getting the initialization path right should be part of #94912, I think. Maybe I should roll back `newReplica` to always creating an uninited replica first, and then quickly fix it up to become inited.

Orthogonally, this PR is more related to #93898 - it consolidates the storage interaction, so we can tweak it in the scope of sep-raft-log work. This PR, I think, also gives you some more flexibility in `Store.Start`: you can do all the IO there, then create all replicas in memory without hidden assumptions about how replica creation path further interacts with the storage (+reuse the invariant checks).

Reply:
I'm fine with this PR, especially since both of us independently (didn't see your comment above for some reason where you had very similar reasoning!) arrived at the conclusion that this is OK. I was just trying to see what the next steps would be.

Looks like the plan is to revisit this in #94912 and move "everything" where it belongs, that LGTM. I'll review the new commits now.

Reply:
Consider leaving a comment on these interim states that explains why it's ok for now and what we're planning to do, etc